### PR TITLE
Add `MapboxNavigationApp` functions to check if `isSetup` and get registered observers

### PR DESCRIPTION
**Pull request by kmadsen**
### Description

investigating https://github.com/mapbox/mapbox-navigation-android/issues/5349
resolves https://github.com/mapbox/mapbox-navigation-android/issues/5350

[Wrote it up in an doc](https://docs.google.com/document/d/1u7T_tuTBK2iLHr6yhYQOSjb7ldVDy6rDuiz0kQCMn7I/edit?usp=sharing) that gives more details while investigating solutions

Considering the ticket, [Make each UI component a standalone, lifecycle-aware object](https://github.com/mapbox/mapbox-navigation-android/issues/5350).
We can use `MapboxNavigationObserver` with Lifecycle extensions, to create standalone lifecycle-aware objects. This pull request shows a handful of examples in the QA app that follow this pattern. 

### Proposal

This pull request, does not add much to the sdk. It adds two functions
``` kotlin
object MapboxNavigationApp
    fun isSetup(): Boolean
    fun <T : MapboxNavigationObserver> getObserver(kClass: KClass<T>): T
```
Considering that MapboxNavigationApp is a singleton, this means that it provides a mechanism for sharing objects as if they are also singletons. `MapboxNavigationApp.getObserver` will crash if the `MapboxNavigationObserver` is not registered through `MapboxNavigationApp.registerObserver`. You can guarantee that your getObserver will never crash if you attach your observers to lifecycles, so this pull request also introduces a few extensions:

``` kotlin
fun <T : MapboxNavigationObserver> LifecycleOwner.attachCreated(vararg observers: T)
fun <T : MapboxNavigationObserver> LifecycleOwner.attachStarted(vararg observers: T)
fun <T : MapboxNavigationObserver> LifecycleOwner.attachResumed(vararg observers: T)
```

This allows you to organize your `MapboxNavigationObserver` implementations and easily hook them into your LifecycleOwners. So this block of code will give the same behavior from a Fragment, Activity, or our NavigationView

``` kotlin
attachCreated(
    DropInLocationCoordinator(),
    DropInNavigationCameraState(binding.mapView),
    DropInReplayComponent(),
)
attachResumed(
    DropInLocationPuck(binding.mapView),
    DropInRoutesInteractor(binding.mapView),
    DropInNavigationCamera(this, binding.mapView),
    DropInRecenterButton(this, binding.recenter),
    DropInContinuousRoutes(),
    DropInActiveGuidance(binding.stop, binding.maneuverView, binding.tripProgressView),
    DropInStartReplayButton(binding.startNavigation),
)
```

Another option to this, is to share the dependencies directly. WIthout requiring a call to getObserver. And then the responsibility for managing these dependencies would be addressed in https://github.com/mapbox/mapbox-navigation-android/issues/5349
``` kotlin
val locationCoordinator = DropInLocationCoordinator()
attachCreated(locationCoordinator)
attachResumed(
    DropInLocationPuck(binding.mapView, locationCoordinator),
    DropInRoutesInteractor(binding.mapView),
    DropInNavigationCamera(this, binding.mapView, locationCoordinator),
    DropInRecenterButton(this, binding.recenter),
    DropInContinuousRoutes(),
)
```

In dropin-ui we have a coroutineScope that allows us to create flowables and collect them `onAttached`
``` kotlin
open class DropInComponent : MapboxNavigationObserver {
    lateinit var coroutineScope: CoroutineScope
```

### Alternative DefaultLifecycleOwner approach

This is still being considered as a solution for https://github.com/mapbox/mapbox-navigation-android/issues/5350. For example, the examples repository has a [AppNavigationCamera](https://github.com/mapbox/mapbox-navigation-android-examples/blob/c50ea353ab49559b7917851bf0a3c37adb4d1553/android-auto-app/src/main/java/com/mapbox/navigation/examples/androidauto/app/navigation/AppNavigationCamera.kt#L28-L34) that can be easily attached onto a lifecycle owner.
``` kotlin
lifecycle.addObserver(AppNavigationCamera(binding.mapView, CameraMode.FOLLOWING))
```

One issue here is `removeObserver` does not have a destruction event. This means, if you `lifecycle.removeObserver`, you still need to clean up that object.
``` kotlin
lifecycle.removebserver(appNavigationCamera)
appNavigationCamera.onDetached()
```

This issue does not exist for `MapboxNavigationObserver`. `MapboxNavigationApp.unregisterObserver(observer)` will trigger `onDetached`. 
```
MapboxNavigationApp.unregisterObserver(observer)
// No extra call, the observer is responsible for implementing onDetached and will be automatically cleaned up.
```

### CODE REVIEW

File: libnavigation-core/src/main/java/com/mapbox/navigation/core/lifecycle/MapboxNavigationApp.kt [hunk 1]
```diff
@@ -131,6 +140,15 @@ object MapboxNavigationApp {
         mapboxNavigationAppDelegate.unregisterObserver(mapboxNavigationObserver)
```
Comment (kmadsen) [inline libnavigation-core/src/main/java/com/mapbox/navigation/core/lifecycle/MapboxNavigationApp.kt:140]: Out of date comment here, but making a note. Was attempting to build this type of function

``` kotlin
   fun <T : MapboxNavigationObserver> registerObserver(lifecycleOwner: LifecycleOwner, observer: MapboxNavigationObserver) {
        return mapboxNavigationAppDelegate.getObserver(kClass)
    }
```

It turned out to be complicated. Maintaining a map of observers, checking on every event if it is supposed to register or unregister the observer. 

There is also an issue with the interface. It's like it should be
``` kotlin
    fun <T : MapboxNavigationObserver> registerObserver(lifecycleOwner: LifecycleOwner, kClass: KClass<T>) {
        return mapboxNavigationAppDelegate.getObserver(kClass)
    }
```

But then we also need to create a factory provider. So I'm finding this solution to be a bit complicated

**Comment by kmadsen** [comment]
Progress update. There's two Activities

1. DropInNavigationViewActivity. An empty activity that has a DropInNavigationView defined in xml
1. AppLifecycleActivity. Which has an ActiveGuidanceFragment. And hosts an DropInActiveGuidance

The examples are attempting to demonstrate the portability and flexibility of the components. This video shows the two simple activities, working with drop in components

https://user-images.githubusercontent.com/3021882/151084885-ab80c818-7a35-49bd-a999-50a761b2bb1e.mp4

File: CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,9 @@ Mapbox welcomes participation and contributions from everyone.
 
 ## Unreleased 
 - Added options to the MapboxRouteLineOptions class to control the icon anchor and icon offset for the route line related waypoint icons including the origin and destination points. [#5409](https://github.com/mapbox/mapbox-navigation-android/pull/5409)
+## Unreleased
```
Comment (VysotskiVadim) [inline CHANGELOG.md:7]: redundant header

**Review by abhishek1508 — APPROVED** [review]

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/MapboxNavigationViewModel.kt [hunk 3]
```diff
@@ -33,7 +33,7 @@ import kotlinx.coroutines.flow.MutableSharedFlow
 import kotlinx.coroutines.launch
 
 @OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
-internal class MapboxNavigationViewModel : ViewModel() {
```
Comment (abhishek1508) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/MapboxNavigationViewModel.kt:36]: Why did we remove `internal`? We should not allow access to any View model to outside developers

**Pull request merged**